src/hardware/driver/relay.rs

```diff
@@ -0,0 +1,256 @@
+///! Driver relays driver
+///!
+///! There are 2 relays at the driver output:
+///!    - one that shorts the output to ground
+///!    - one that connects the current source/sink to the output
+///!
+///! See [Relay] documentation for details about relay control.
+///!
+///! The relays are controlled via an MCP23008 I2C io expander.
+use super::hal::rcc;
+use core::fmt::Debug;
+use embedded_hal::blocking::{
+    delay::DelayUs,
+    i2c::{Write, WriteRead},
+};
+use mcp230xx::{Level, Mcp23008, Mcp230xx};
+
+use super::Channel;
+use smlang::statemachine;
+
+// Driver output relays pins.
+#[allow(non_camel_case_types)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+enum RelayPin {
+    LN_K1_EN_N,
+    LN_K1_EN,
+    LN_K0_D,
+    LN_K0_CP,
+    HP_K1_EN_N,
+    HP_K1_EN,
+    HP_K0_D,
+    HP_K0_CP,
+}
+
+impl From<RelayPin> for Mcp23008 {
+    fn from(pin: RelayPin) -> Self {
+        match pin {
+            RelayPin::LN_K1_EN_N => Self::P0,
+            RelayPin::LN_K1_EN => Self::P1,
+            RelayPin::LN_K0_D => Self::P2,
+            RelayPin::LN_K0_CP => Self::P3,
+            RelayPin::HP_K1_EN_N => Self::P4,
+            RelayPin::HP_K1_EN => Self::P5,
+            RelayPin::HP_K0_D => Self::P6,
+            RelayPin::HP_K0_CP => Self::P7,
+        }
+    }
+}
+
+// small helper to lock the mutex
+// maybe todo: Pass out an error if in use. Not sure how to get that out of the state machine though..
+fn get_mcp<I2C>(
```

Comment:
It's a one-liner. Why not inline this?

Reply:
oops, this was more code at some point..